File: p2pnet/android/app/src/main/java/com/example/p2pnet/data/remote/WsClient.kt
package com.example.p2pnet.data.remote

import android.os.Handler
import android.os.Looper
import com.example.p2pnet.data.local.LocalPrefs
import okhttp3.*
import org.json.JSONObject
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.Inet4Address
import java.net.InetAddress
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class WsClient(
    private val localPrefs: LocalPrefs? = null
) {
    interface Listener {
        fun onConnected()
        fun onDisconnected()
        fun onRawMessage(text: String)
        fun onSend(text: String)
        fun onLoginSuccess(username: String)
        fun onLoginFailed(message: String)
        fun onError(message: String)
        fun onUdpSend(text: String)
        fun onUdpRecv(text: String)
        fun onHelloDone(info: PeerInfo?, sock: DatagramSocket?, peerIp: String, peerPort: Int, mode: String)
    }

    data class PeerInfo(
        val name: String,
        val peerIp: String,
        val peerPort: Int,
        val myIp: String,
        val myPort: Int,
        val myLocalPort: Int
    )

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private var ws: WebSocket? = null
    var listener: Listener? = null
    private var serverIp = ""
    private var serverUdpPort = 0
    private var serverHost = ""

    // Login state
    private var loginUsername: String = ""
    private var loginPassword: String = ""
    var confirmedUsername: String = ""
    private var sessionKey: ByteArray? = null
    private var pendingSalt: String = ""
    private var pendingChallenge: String = ""
    private var pendingPwHash: String = ""

    // UDP hello 共享状态（主线程创建socket，hello线程写，主线程读）
    private var _pendingPeerInfo: PeerInfo? = null
    // "udp" or "wg"：标识当前 hello 线程是为 UDP tab 还是 WireGuard tab
    private var _helloMode: String = "udp"
    private var _helloPeerIp: String = ""
    private var _helloPeerPort: Int = 0
    // 中断标志：服务器发 thisisyourpeer 时置 true，hello 线程检查到这个标志就立即停止发包并退出
    private val stopFlag = AtomicBoolean(false)
    private var helloThread: Thread? = null

    fun connect(url: String) {
        val request = Request.Builder().url(url).build()
        ws = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                listener?.onConnected()
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                listener?.onError(t.message ?: "connection failed")
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                listener?.onDisconnected()
            }
        })
    }

    private fun handleMessage(text: String) {
        try {
            val obj = JSONObject(text)
            val type = obj.optString("type")
            listener?.onRawMessage(text)

            when (type) {
                "login_failed" -> {
                    listener?.onLoginFailed(obj.optString("message"))
                }
                "send_udp_to_server" -> {
                    serverIp = obj.optString("server_ip", "")
                    serverUdpPort = obj.optInt("udpport", 0)
                    listener?.onSend("send_udp_to_server")
                    android.util.Log.e("UDP", "send_udp_to_server received! serverIp=$serverIp serverUdpPort=$serverUdpPort")
                    startUdpHello()
                }
                "thisisyourpeer_udp" -> {
                    val info = PeerInfo(
                        name = obj.optString("name", ""),
                        peerIp = obj.optString("ip", ""),
                        peerPort = obj.optInt("port", 0),
                        myIp = obj.optString("my_ip", ""),
                        myPort = obj.optInt("my_port", 0),
                        myLocalPort = 0  // 待 hello 线程填入
                    )
                    _pendingPeerInfo = info
                    _helloPeerIp = info.peerIp
                    _helloPeerPort = info.peerPort
                    listener?.onUdpRecv("收到 thisisyourpeer_udp，设置停止标志")
                    stopFlag.set(true)
                }
                "challenge" -> {
                    val challenge = obj.optString("challenge", "")
                    val salt = obj.optString("salt", "")
                    pendingSalt = salt
                    pendingChallenge = challenge
                    val pwHash = sha256(loginPassword + salt)
                    pendingPwHash = pwHash
                    val response = hmacSha256(hexDecode(pwHash), hexDecode(challenge))
                    sendJson(JSONObject().apply {
                        put("type", "login")
                        put("username", loginUsername)
                        put("response", response)
                    })
                    loginPassword = ""
                }
                "login_ok" -> {
                    confirmedUsername = obj.optString("username", "")
                    val skIkm = hexDecode(pendingPwHash)
                    sessionKey = hkdfSha256(skIkm, skIkm, hexDecode(pendingChallenge))
                    android.util.Log.e("UDP", "session_key derived: ${sessionKey?.let { binasciiHexlify(it) }}")
                    listener?.onLoginSuccess(confirmedUsername)
                }
            }
        } catch (e: Exception) {
            listener?.onError("parse error: ${e.message}")
        }
    }

    private fun startUdpHello() {
        // 捕获当前 mode（sendP2pUdp 或 sendWghelp 设置的），hello 线程结束时用它决定路由
        val mode = _helloMode
        listener?.onUdpSend("startUdpHello ENTRY serverIp=$serverIp serverUdpPort=$serverUdpPort mode=$mode")
        if (serverIp.isEmpty() && serverHost.isNotEmpty()) {
            try {
                serverIp = InetAddress.getByName(serverHost).hostAddress ?: ""
                listener?.onUdpSend("serverIp fallback resolved: $serverIp from $serverHost")
            } catch (e: Exception) {
                listener?.onUdpSend("serverIp fallback failed: ${e.message}")
            }
        }
        if (serverIp.isEmpty() || serverUdpPort == 0) {
            listener?.onUdpSend("startUdpHello EARLY RETURN! serverIp=$serverIp serverUdpPort=$serverUdpPort")
            return
        }

        // 主线程创建 socket，然后直接返回，不阻塞 WebSocket 线程
        var sock: DatagramSocket? = null
        try {
            val ipv4Addr = Inet4Address.getByName("0.0.0.0")
            sock = DatagramSocket(null as java.net.InetSocketAddress?)
            sock.bind(java.net.InetSocketAddress(ipv4Addr, 0))
            listener?.onUdpSend("UDP hello socket 绑定类型=${sock.localAddress.javaClass.name} 本地端口=${sock.localPort} 目标=$serverIp:$serverUdpPort")
            val localPort = sock.localPort
            // 重置共享状态
            stopFlag.set(false)
            _pendingPeerInfo = null

            // 启动 hello 临时线程，不 join，让 WebSocket 线程立即返回
            helloThread = Thread {
                try {
                    runHello(sock!!, localPort)
                } catch (e: Exception) {
                    listener?.onError("hello thread error: ${e.message}")
                } finally {
                    // hello 线程结束后，根据结果决定关 socket 还是传给新 tab
                    val info = _pendingPeerInfo
                    Handler(Looper.getMainLooper()).post {
                        if (info != null) {
                            listener?.onUdpSend("hello 线程结束，收到 peer info，拉起新 tab")
                            listener?.onHelloDone(info, sock, _helloPeerIp, _helloPeerPort, mode)
                        } else {
                            listener?.onUdpSend("hello 线程结束，未收到 peer info，关闭 socket")
                            try { sock.close() } catch (_: Exception) {}
                            listener?.onHelloDone(null, null, "", 0, mode)
                        }
                    }
                }
            }
            helloThread?.start()
        } catch (e: Exception) {
            listener?.onError("startUdpHello error: ${e.message}")
            try { sock?.close() } catch (_: Exception) {}
        }
    }

    /**
     * Hello 临时线程：只发包，不创建也不关闭 socket。
     * 通过 stopFlag AtomicBoolean 被 thisisyourpeer_udp 中断。
     */
    private fun runHello(sock: DatagramSocket, localPort: Int) {
        val addr = InetAddress.getByName(serverIp)

        // Phase 1: burst 15 个，30ms 间隔
        for (i in 0 until 15) {
            val payload = buildP2pUdpPayload()
            val data = payload.toString().toByteArray()
            val pkt = DatagramPacket(data, data.size, addr, serverUdpPort)
            sock.send(pkt)
            listener?.onUdpSend("UDP burst[$i] → $payload")
            Thread.sleep(30)
        }
        listener?.onUdpSend("burst 15个发完，进入维持阶段")

        // Phase 2: 每秒发1个维持包，可被 stopFlag 中断
        var seq = 0
        val startTime = System.currentTimeMillis()
        while (!stopFlag.get()) {
            if (System.currentTimeMillis() - startTime > 10_000) {
                listener?.onUdpSend("UDP hello 维持 10s 无响应，主动放弃")
                return
            }
            Thread.sleep(1000)
            if (stopFlag.get()) break
            seq++

            val payload = buildP2pUdpPayload()
            val data = payload.toString().toByteArray()
            val pkt = DatagramPacket(data, data.size, addr, serverUdpPort)
            sock.send(pkt)
            listener?.onUdpSend("UDP keep-alive[$seq] → $payload")
        }

        // stopFlag 被设置（收到了 thisisyourpeer）
        if (_pendingPeerInfo != null) {
            _pendingPeerInfo = _pendingPeerInfo!!.copy(myLocalPort = localPort)
            listener?.onUdpSend("hello 线程退出，收到 peer info: ${_pendingPeerInfo?.peerIp}:${_pendingPeerInfo?.peerPort}")
        }

        // 被 stopFlag 停止（收到了 thisisyourpeer）
        if (_pendingPeerInfo != null) {
            // 填入本地端口
            _pendingPeerInfo = _pendingPeerInfo!!.copy(myLocalPort = localPort)
            listener?.onUdpSend("hello 线程被中断，收到 peer info: ${_pendingPeerInfo?.peerIp}:${_pendingPeerInfo?.peerPort}")
        }
    }

    private fun buildP2pUdpPayload(): JSONObject {
        val sig = sessionKey?.let { hmacSha256Hex(it, "ping".toByteArray()) }
        return JSONObject().apply {
            put("type", "p2pudp_hello")
            put("username", confirmedUsername)
            sig?.let { put("signature", it) }
        }
    }

    fun resetUdpState() {
        stopFlag.set(true)
        helloThread?.interrupt()
        helloThread = null
    }

    fun sendP2pUdp(target: String) {
        _helloMode = "udp"
        sendJson(JSONObject().put("type", "p2pudp").put("target", target))
    }

    fun sendP2pTcp(target: String) {
        sendJson(JSONObject().put("type", "p2ptcp").put("target", target))
    }

    fun sendWghelp(target: String) {
        _helloMode = "wg"
        sendJson(JSONObject().put("type", "wghelp").put("target", target))
    }

    fun sendList() {
        sendJson(JSONObject().put("type", "list"))
    }

    fun disconnectOnly() {
        resetUdpState()
        ws?.close(1000, "bye")
        ws = null
    }

    fun login(useWss: Boolean, serverHost: String, serverPort: Int, username: String, password: String) {
        val proto = if (useWss) "wss" else "ws"
        connect("$proto://$serverHost:$serverPort/")
        this.serverHost = serverHost
        loginUsername = username
        loginPassword = password
        sendJson(JSONObject().apply {
            put("type", "login")
            put("username", username)
        })
    }

    private fun sendJson(obj: JSONObject) {
        listener?.onSend(obj.toString())
        ws?.send(obj.toString())
    }

    // ---- Python-compatible crypto utils ----
    private fun sha256(data: String): String {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(data.toByteArray())
        return hash.joinToString("") { "%02x".format(it) }
    }

    private fun hmacSha256(key: ByteArray, data: ByteArray): String {
        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(javax.crypto.spec.SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data).joinToString("") { "%02x".format(it) }
    }

    private fun hmacSha256Hex(key: ByteArray, data: ByteArray): String {
        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(javax.crypto.spec.SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data).joinToString("") { "%02x".format(it) }
    }

    private fun hexDecode(hex: String): ByteArray {
        val result = ByteArray(hex.length / 2)
        for (i in hex.indices step 2) {
            result[i / 2] = ((Character.digit(hex[i], 16) shl 4) + Character.digit(hex[i + 1], 16)).toByte()
        }
        return result
    }

    private fun binasciiHexlify(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }

    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray): ByteArray {
        val prkMac = javax.crypto.Mac.getInstance("HmacSHA256")
        prkMac.init(javax.crypto.spec.SecretKeySpec(salt, "HmacSHA256"))
        val prk = prkMac.doFinal(ikm)
        val n = 32 / 32
        var t = ByteArray(0)
        var okm = ByteArray(0)
        var i = 1
        while (okm.size < 32) {
            val input = t + info + byteArrayOf(i.toByte())
            val mac = javax.crypto.Mac.getInstance("HmacSHA256")
            mac.init(javax.crypto.spec.SecretKeySpec(prk, "HmacSHA256"))
            t = mac.doFinal(input)
            okm += t
            i++
        }
        return okm.copyOf(32)
    }
}

File: p2pnet/ios/p2pnet/Data/Remote/WsClient.swift
import Foundation

// MARK: - PeerInfo

struct PeerInfo: Equatable {
    let name: String
    let peerIp: String
    let peerPort: Int
    let myIp: String
    let myPort: Int
    var myLocalPort: Int
}

// MARK: - WsClientListener

protocol WsClientListener: AnyObject {
    func onMessage(_ text: String)
    func onConnected()
    func onDisconnected()
    func onRecv(_ text: String)
    func onSend(_ text: String)
    func onLoginSuccess(_ username: String)
    func onLoginFailed(_ message: String)
    func onError(_ message: String)
    func onUdpSend(_ text: String)
    func onUdpRecv(_ text: String)
    func onHelloDone(_ info: PeerInfo?, _ sock: Int32?, _ peerIp: String, _ peerPort: Int, _ mode: String)
}

// MARK: - WsClient using URLSessionWebSocketTask (Foundation)

class WsClient: NSObject, URLSessionWebSocketDelegate {
    private var session: URLSession?
    private var wsTask: URLSessionWebSocketTask?
    weak var listener: WsClientListener?

    private var serverHost = ""
    private var serverPort = 0
    private var useWss = false

    private var loginUsername = ""
    private var loginPassword = ""
    var confirmedUsername = ""
    private var sessionKey: Data?
    private var pendingSalt = ""
    private var pendingChallenge = ""
    private var pendingPwHash = ""

    private var _pendingPeerInfo: PeerInfo?
    private var _helloMode = "udp"
    private var _helloPeerIp = ""
    private var _helloPeerPort = 0
    private var stopFlag = false
    private var helloQueue: DispatchQueue?

    // UDP socket fd (for P2P hello)
    private var udpSockFd: Int32 = -1
    private var serverIp = ""
    private var serverUdpPort = 0

    // Track connection state
    private var isReceiving = false
    private var isConnected = false

    // Pending messages to send after connection opens
    private var pendingMessages: [String] = []

    override init() {
        super.init()
    }

    func connect(useWss: Bool, host: String, port: Int) {
        self.serverHost = host
        self.serverPort = port
        self.useWss = useWss
        self.isConnected = false
        self.pendingMessages = []

        let proto = useWss ? "wss" : "ws"
        let urlStr = "\(proto)://\(host):\(port)/"
        print("[WsClient] connect() listener=\(String(format: "%p", unsafeBitCast(listener as AnyObject, to: Int.self))) listener?.onMessage: \(listener != nil ? "YES" : "NO")")
        listener?.onMessage("连接 \(urlStr)")
        print("[WsClient] connect() url=\(urlStr)")

        guard let url = URL(string: urlStr) else {
            print("[WsClient] connect() FAIL: invalid url")
            listener?.onMessage("无效的 URL: \(urlStr)")
            return
        }

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 0

        session = URLSession(configuration: config, delegate: self, delegateQueue: nil)
        wsTask = session?.webSocketTask(with: url)
        wsTask?.resume()
        listener?.onMessage("URLSession 已创建，wsTask.resume() 已调用")
    }

    // MARK: - URLSessionWebSocketDelegate

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol proto: String?) {
        print("[WsClient] didOpenWithProtocol: \(proto ?? "nil")")
        isConnected = true
        listener?.onMessage("WebSocket 连接已打开，协议: \(proto ?? "nil")")
        listener?.onConnected()

        // Flush pending messages
        for text in pendingMessages {
            print("[WsClient] didOpen flush pending: \(text)")
            doSend(text)
        }
        pendingMessages = []

        startReceiveLoop()
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: Int, reason: Data?) {
        print("[WsClient] didCloseWith closeCode=\(closeCode)")
        isConnected = false
        isReceiving = false
        listener?.onMessage("WebSocket 连接关闭 closeCode=\(closeCode)")
        listener?.onDisconnected()
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        print("[WsClient] didCompleteWithError: \(error?.localizedDescription ?? "nil")")
        isConnected = false
        isReceiving = false
        if let error = error {
            let nsErr = error as NSError
            // ATS / NSURLErrorDomain errors
            if nsErr.domain == "NSURLErrorDomain" {
                let msg: String
                switch nsErr.code {
                case -1022:
                    msg = "ATS拒绝: App Transport Security 阻止非安全连接 (ws://). 需在 Info.plist 添加 NSAppTransportSecurity 配置，或使用 wss://"
                case -1001:
                    msg = "连接超时"
                case -1003:
                    msg = "找不到服务器: \(serverHost):\(serverPort)"
                case -1004:
                    msg = "无法连接服务器"
                case -1005:
                    msg = "网络连接丢失"
                case -1200:
                    msg = "TLS/SSL错误"
                default:
                    msg = "连接错误 [\(nsErr.code)]: \(error.localizedDescription)"
                }
                listener?.onMessage(msg)
            } else {
                print("[WsClient] didCompleteWithError listener?.onMessage was NOT called (no error)")
                listener?.onMessage(error.localizedDescription)
            }
        }
    }

    private func startReceiveLoop() {
        guard !isReceiving else {
            listener?.onMessage("startReceiveLoop: 已在接收中，跳过")
            return
        }
        isReceiving = true

        wsTask?.receive { [weak self] result in
            guard let self = self else { return }
            self.isReceiving = false

            switch result {
            case .success(let msg):
                print("[WsClient] receive success: \(msg)")
                switch msg {
                case .string(let text):
                    listener?.onRecv(text)
                    self.handleMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        listener?.onRecv(text)
                        self.handleMessage(text)
                    }
                @unknown default:
                    break
                }
                if self.wsTask != nil && self.isConnected {
                    self.startReceiveLoop()
                }

            case .failure(let err):
                let nsErr = err as NSError
                print("[WsClient] receive failure: code=\(nsErr.code) msg=\(err.localizedDescription)")
                if nsErr.code == 57 || nsErr.code == 1 {
                    self.listener?.onDisconnected()
                } else if nsErr.code == -1001 {
                    self.listener?.onMessage("接收超时")
                } else if nsErr.code == -1005 {
                    self.listener?.onMessage("网络连接丢失")
                } else {
                    self.listener?.onMessage("接收错误 [\(nsErr.code)]: \(err.localizedDescription)")
                }
            }
        }
    }

    func disconnectOnly() {
        print("[WsClient] disconnectOnly()")
        resetUdpState()
        isReceiving = false
        isConnected = false
        wsTask?.cancel(with: .goingAway, reason: nil)
        wsTask = nil
        session?.invalidateAndCancel()
        session = nil
        listener?.onMessage("已断开连接")
        listener?.onDisconnected()
    }

    private func handleMessage(_ text: String) {
        print("[WsClient] handleMessage: \(text)")

        guard let data = text.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = obj["type"] as? String else { return }

        print("[WsClient] handleMessage type=\(type)")

        switch type {
        case "login_failed":
            listener?.onLoginFailed(obj["message"] as? String ?? "")

        case "send_udp_to_server":
            serverIp = obj["server_ip"] as? String ?? ""
            serverUdpPort = obj["udpport"] as? Int ?? 0
            listener?.onSend("send_udp_to_server")
            startUdpHello()

        case "thisisyourpeer_udp":
            let info = PeerInfo(
                name: obj["name"] as? String ?? "",
                peerIp: obj["ip"] as? String ?? "",
                peerPort: obj["port"] as? Int ?? 0,
                myIp: obj["my_ip"] as? String ?? "",
                myPort: obj["my_port"] as? Int ?? 0,
                myLocalPort: 0
            )
            _pendingPeerInfo = info
            _helloPeerIp = info.peerIp
            _helloPeerPort = info.peerPort
            stopFlag = true
            listener?.onUdpRecv("收到 thisisyourpeer_udp，设置停止标志")

        case "challenge":
            print("[WsClient] got challenge, computing response")
            pendingChallenge = obj["challenge"] as? String ?? ""
            pendingSalt = obj["salt"] as? String ?? ""
            let pwHash = Crypto.sha256(loginPassword + pendingSalt)
            pendingPwHash = pwHash
            let response = Crypto.computeAuthResponse(password: loginPassword, salt: pendingSalt, challenge: pendingChallenge)
            sendJson(["type": "login", "username": loginUsername, "response": response])
            loginPassword = ""

        case "login_ok":
            confirmedUsername = obj["username"] as? String ?? ""
            print("[WsClient] login_ok username=\(confirmedUsername)")
            listener?.onLoginSuccess(confirmedUsername)

        default:
            break
        }
    }

    private func startUdpHello() {
        print("[WsClient] startUdpHello ENTRY serverIp=\(serverIp) serverUdpPort=\(serverUdpPort) mode=\(_helloMode)")
        if serverIp.isEmpty || serverUdpPort == 0 {
            listener?.onUdpSend("startUdpHello EARLY RETURN! serverIp=\(serverIp) serverUdpPort=\(serverUdpPort)")
            return
        }

        let sockFd = Darwin.socket(AF_INET, SOCK_DGRAM, 0)
        if sockFd < 0 {
            listener?.onMessage("UDP socket 创建失败")
            return
        }
        listener?.onMessage("UDP socket 创建成功 fd=\(sockFd)")
        udpSockFd = sockFd

        var addr = sockaddr_in()
        addr.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_addr.s_addr = 0
        addr.sin_port = 0

        let bindResult = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                Darwin.bind(sockFd, sockaddrPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }

        if bindResult < 0 {
            listener?.onMessage("UDP bind 失败")
            Darwin.close(sockFd)
            udpSockFd = -1
            return
        }

        listener?.onMessage("UDP bind 成功")
        stopFlag = false
        _pendingPeerInfo = nil

        helloQueue = DispatchQueue(label: "p2pnet.udp.hello", qos: .userInitiated)
        helloQueue?.async { [weak self] in
            guard let self = self else { return }
            let lp = self.getLocalUdpPort(sockFd)
            self.runHello(sockFd, lp)
        }
    }

    private func runHello(_ sockFd: Int32, _ localPort: Int) {
        guard let addrCStr = serverIp.withCString({ strdup($0) }) else { return }
        defer { free(addrCStr) }

        var destAddr = sockaddr_in()
        destAddr.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        destAddr.sin_family = sa_family_t(AF_INET)
        destAddr.sin_port = UInt16(serverUdpPort).bigEndian
        inet_pton(AF_INET, addrCStr, &destAddr.sin_addr)

        func sendToPeer(_ data: Data) {
            data.withUnsafeBytes { dataPtr in
                withUnsafePointer(to: &destAddr) { destPtr in
                    destPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                        let sent = sendto(sockFd, dataPtr.baseAddress, data.count, 0, sockaddrPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
                        if sent < 0 {
                            listener?.onMessage("UDP sendto 失败 errno=\(errno)")
                        }
                    }
                }
            }
        }

        listener?.onMessage("UDP hello 开始，发送 15 个 burst")
        for i in 0..<15 {
            let payload = buildP2pUdpPayload()
            if let data = payload.data(using: .utf8) {
                sendToPeer(data)
            }
            listener?.onUdpSend("UDP burst[\(i)] → \(payload)")
            Thread.sleep(forTimeInterval: 0.03)
        }
        listener?.onUdpSend("burst 15个发完，进入维持阶段")

        var seq = 0
        let startTime = Date()
        while !stopFlag {
            if Date().timeIntervalSince(startTime) > 10 {
                listener?.onUdpSend("UDP hello 维持 10s 无响应，主动放弃")
                cleanupSocket(sockFd)
                listener?.onMessage("UDP hello 超时")
                notifyHelloDone(nil, nil)
                return
            }
            Thread.sleep(forTimeInterval: 1)
            if stopFlag { break }
            seq += 1

            let payload = buildP2pUdpPayload()
            if let data = payload.data(using: .utf8) {
                sendToPeer(data)
            }
            listener?.onUdpSend("UDP keep-alive[\(seq)] → \(payload)")
        }

        if var updatedInfo = _pendingPeerInfo {
            updatedInfo = PeerInfo(
                name: updatedInfo.name,
                peerIp: updatedInfo.peerIp,
                peerPort: updatedInfo.peerPort,
                myIp: updatedInfo.myIp,
                myPort: updatedInfo.myPort,
                myLocalPort: localPort
            )
            listener?.onUdpSend("hello 线程被中断，收到 peer info: \(updatedInfo.peerIp):\(updatedInfo.peerPort)")
            cleanupSocket(sockFd)
            notifyHelloDone(updatedInfo, sockFd)
        } else {
            cleanupSocket(sockFd)
            listener?.onMessage("UDP hello 未找到 peer 信息")
            notifyHelloDone(nil, nil)
        }
    }

    private func getLocalUdpPort(_ sockFd: Int32) -> Int {
        var localAddr = sockaddr_in()
        var len = socklen_t(MemoryLayout<sockaddr_in>.size)
        withUnsafeMutablePointer(to: &localAddr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                getsockname(sockFd, sockaddrPtr, &len)
            }
        }
        return Int(localAddr.sin_port).bigEndian
    }

    private func cleanupSocket(_ sockFd: Int32) {
        Darwin.close(sockFd)
        listener?.onMessage("UDP socket 已关闭 fd=\(sockFd)")
        if udpSockFd == sockFd {
            udpSockFd = -1
        }
    }

    private func notifyHelloDone(_ info: PeerInfo?, _ sockFd: Int32?) {
        if let info = info {
            listener?.onMessage("UDP hello 成功，peer=\(info.peerIp):\(info.peerPort)")
        }
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.listener?.onHelloDone(info, sockFd, self._helloPeerIp, self._helloPeerPort, self._helloMode)
        }
    }

    private func buildP2pUdpPayload() -> String {
        var payload: [String: Any] = [
            "type": "p2pudp_hello",
            "username": confirmedUsername
        ]
        if let key = sessionKey {
            let sig = Crypto.hmacSHA256Hex(key: key, data: Data("ping".utf8))
            payload["signature"] = sig
        }
        if let data = try? JSONSerialization.data(withJSONObject: payload),
           let str = String(data: data, encoding: .utf8) {
            return str
        }
        return "{\"type\":\"p2pudp_hello\",\"username\":\"\(confirmedUsername)\"}"
    }

    func resetUdpState() {
        stopFlag = true
        helloQueue?.async { [weak self] in
            if let fd = self?.udpSockFd, fd >= 0 {
                Darwin.close(fd)
                self?.udpSockFd = -1
            }
        }
        helloQueue = nil
    }

    func sendP2pUdp(_ target: String) {
        _helloMode = "udp"
        sendJson(["type": "p2pudp", "target": target])
    }

    func sendP2pTcp(_ target: String) {
        sendJson(["type": "p2ptcp", "target": target])
    }

    func sendWghelp(_ target: String) {
        _helloMode = "wg"
        sendJson(["type": "wghelp", "target": target])
    }

    func sendList() {
        sendJson(["type": "list"])
    }

    func login(useWss: Bool, serverHost: String, serverPort: Int, username: String, password: String) {
        self.serverHost = serverHost
        loginUsername = username
        loginPassword = password
        print("[WsClient] login() username=\(username) useWss=\(useWss) host=\(serverHost):\(serverPort)")
        if isConnected {
            print("[WsClient] login() already connected, sending login message")
            let msg = "{\"type\":\"login\",\"username\":\"\(username)\"}"
            listener?.onSend(msg)
            doSend(msg)
            return
        }
        listener?.onMessage("请先点击连接")
        return
    }

    private func sendJson(_ obj: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: obj),
              let text = String(data: data, encoding: .utf8) else { return }
        print("[WsClient] sendJson: \(text)")
        listener?.onSend(text)

        if isConnected {
            doSend(text)
        } else {
            print("[WsClient] sendJson QUEUED (not connected yet): \(text)")
            pendingMessages.append(text)
        }
    }

    private func doSend(_ text: String) {
        let msg = URLSessionWebSocketTask.Message.string(text)
        wsTask?.send(msg) { [weak self] error in
            if let error = error {
                let nsErr = error as NSError
                if nsErr.code == -1001 {
                    self?.listener?.onMessage("发送超时")
                } else if nsErr.code == -1004 {
                    self?.listener?.onMessage("无法发送: 连接已断开")
                } else {
                    self?.listener?.onMessage("发送错误: \(error.localizedDescription)")
                }
            } else {
                print("[WsClient] doSend SUCCESS: \(text)")
            }
        }
    }
}
